src/components/usePopup.ts

```diff
@@ -1,18 +1,83 @@
 import { useEffect, useState } from "react";
-import { CountryName } from "../constants/Countries";
+import { Countries, CountryName, WorldAverage } from "../constants/Countries";
 
 export type PopupProps = {
-    selectedCountries: Set<CountryName>;
+    selectedCountries: Map<CountryName, number>;
     addSelectedCountry: (country: CountryName) => void;
     removeSelectedCountry: (country: CountryName) => void;
-    refreshAndGetSize: (selectedCountry: Set<CountryName>) => Promise<void>;
+    setCountryPercentage: (country: CountryName, percentage: number) => void;
+    averageSpecificEmissions: number;
+    refreshAndGetSize: () => Promise<void>;
+    error: string | null;
 }
 
 export const usePopup = (): PopupProps => {
     const [transferSize, setTransferSize] = useState(0);
-    const [selectedCountries, setSelectedCountries] = useState<Set<CountryName>>(new Set())
+    const [selectedCountries, setSelectedCountries] = useState<Map<CountryName, number>>(new Map<CountryName, number>())
+    const [averageSpecificEmissions, setAverageSpecificEmissions] = useState(0);
+    const [error, setError] = useState<string | null>(null);
+
+    const setCountryPercentage = (country: CountryName, percentage: number) => {
+        const newMap = new Map(selectedCountries);
+        newMap.set(country, percentage);
+        setSelectedCountries(newMap);
+    }
+
+    const calculateAverageSpecificEmissions = () => {
+        if (selectedCountries.size === 0) {
+            setAverageSpecificEmissions(WorldAverage)
+            return;
+        }
+        let totalPercentage = 0;
+        let average = 0;
+        let noOfMissingPercentages = 0;
```

Comment:
I mention it below but "let" is a bit of a red flag can you think of why?

Reply:
Had a look at the ways you can declare variables in JS, let me know if I got them wrong:

- Var 
  - globally/locally scoped
  - Can be redeclared and reassigned
  - Hoisted to top of scope and assigned undefined 
- Let
  - Block scoped
  - Can be reassigned but not redeclared
  - Hoisted but not initialised (returns Reference Error)
- Const
  - block scoped
  - Cannot be redeclared or reassigned
  - Hoisted but not initialised

I can see the issue with `var` because if declared in a file it exists globally and can pollute the namespace, since `var` can be redeclared and reassigned which can cause unexpected behaviour across different files.

I read that `let` may not be supported on older browsers without a transpiler, and using a transpiler may lead to a larger size? I'm not sure how else I can implement a counter in a for loop without `let`, since `const` cannot be redeclared or reassigned

What were your thoughts?

Reply:
It goes back to that mutatability issue, because let is mutable it can have side effects, making it hard to debug, lead to more issues and therefore more bugs because the variable might change type completely. Therefore where possible it is best to use const.

There are ways to loop over a const with a counter using a `.map`